File: rootLeaf.py
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        ans = list()
        def helper(node,num,nums):
            if not node:
                return nums.append("0")
            
            if not node.left and not node.right:
                num += str(node.val)
                nums.append(num)
                return
            num += str(node.val)
            
            helper(node.left,num,nums)
            helper(node.right,num,nums)
            
        helper(root,"",ans) 
        return int(reduce(lambda x,y:int(x)+int(y),ans))

File: root_leaf.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int sum=0;
    void helper(TreeNode root,int current) {
        if(root == null) {
            return;
        }
        current =  current*10+root.val;
        if(root.left == null &&  root.right == null) {
            sum = sum+current;
        }
        helper(root.left,current);
        helper(root.right,current);
    }
    public int sumNumbers(TreeNode root) {
        helper(root,0);
        return sum;
    }
}